File: PostalCode/ViewController/SelectViewController.m
//
//  SelectViewController.m
//  PostalCode
//
//  Created by OCHIISHI Koichiro on 10/22/13.
//  Copyright (c) 2013 OCHIISHI Koichiro. All rights reserved.
//

#import "SelectViewController.h"

@interface SelectViewController ()

@property (nonatomic, strong) NSMutableArray *objects;
@property (nonatomic, strong) NSMutableArray *sectionIndexTitles; // 4文字制限

@end

@implementation SelectViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.objects = [NSMutableArray new];
    self.sectionIndexTitles = [NSMutableArray new];
    self.tableView.sectionIndexTrackingBackgroundColor = [UIColor colorWithRed:206.f/255.f green:203.f/255.f blue:198.f/255.f alpha:.2f];
    
    if (self.selectedAddress != SelectedAddressState) {
        self.navigationItem.leftBarButtonItem = nil;
    }
    
    [(MyNavigationController *)self.navigationController startLoading];
    
    dispatch_queue_t q_global = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    dispatch_queue_t q_main = dispatch_get_main_queue();
    
    dispatch_async(q_global, ^{
        
        if (self.selectedAddress == SelectedAddressState) {
            self.objects = [[PostalCodeRepository sharedRepository] getStates];
            self.sectionIndexTitles = [[PostalCodeRepository sharedRepository] getStateSectionIndexTitles];
        } else if (self.selectedAddress == SelectedAddressCityTown) {
            self.objects = [[PostalCodeRepository sharedRepository] getCityTownsByState:self.selectedState];
            for (PostalCodeModel *model in self.objects) {
                [self.sectionIndexTitles addObject:[self stringToThreeWithString:model.cityTownK]];
            }
        } else if (self.selectedAddress == SelectedAddressStreet) {
            self.objects = [[PostalCodeRepository sharedRepository] getStreetsByState:self.selectedState byCityAndTown:self.selectedCityTown];
            for (NSMutableArray *models in self.objects) {
                PostalCodeModel *model = [models lastObject];
                if (model.streetH.length) {
                    NSString *string = [model.streetH substringToIndex:1];
                    if (string.length) {
                        [self.sectionIndexTitles addObject:string];
                    }
                }
            }
        }
        dispatch_async(q_main, ^{
            [(MyNavigationController *)self.navigationController stopLoading];
            [self.tableView reloadData];
        });
    });
}

- (void)viewWillDisappear:(BOOL)animated
{
    // 「読み込み中」が表示された状態で「戻る」が押された場合に、「読み込み中」が表示され続けてしまう問題を修正
    // http://ameblo.jp/satoko-ohtsuki/entry-11369448573.html
    if ([self.navigationController.viewControllers indexOfObject:self] == NSNotFound) {
        [(MyNavigationController *)self.navigationController stopLoading];
    }
    [super viewWillDisappear:animated];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark -

- (NSString *)stringToThreeWithString:(NSString *)string
{
    return (string.length > 3) ? [string substringToIndex:3] : string;
}

- (NSString *)textLabelText:(NSIndexPath *)indexPath
{
    switch (self.selectedAddress) {
        case SelectedAddressState:
            return self.objects[indexPath.section][indexPath.row];
        case SelectedAddressCityTown: {
            PostalCodeModel *model = self.objects[indexPath.section];
            return model.cityTownK;
        }
        case SelectedAddressStreet: {
            PostalCodeModel *model = self.objects[indexPath.section][indexPath.row];
            return (model.streetK.length) ? model.streetK : model.cityTownK;
        }
        default:
            return @"";
    }
}

- (NSString *)detailTextLabelText:(NSIndexPath *)indexPath
{
    switch (self.selectedAddress) {
        case SelectedAddressCityTown: {
            PostalCodeModel *model = self.objects[indexPath.section];
            return model.cityTownH;
        }
        case SelectedAddressStreet: {
            PostalCodeModel *model = self.objects[indexPath.section][indexPath.row];
            return (model.streetK.length) ? model.streetH : model.cityTownH;
        }
        default:
            return @"";
    }
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.objects.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if (self.selectedAddress == SelectedAddressCityTown) {
        return 1;
    }
    return [self.objects[section] count];
}

- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView
{
    return [self.sectionIndexTitles copy];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if (self.selectedAddress == SelectedAddressState) {
        return self.sectionIndexTitles[section];
    }
    return nil;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [DynamicTypeHelper heightWithStyle:UITableViewCellStyleSubtitle
                                          text:[self textLabelText:indexPath]
                                    detailText:[self detailTextLabelText:indexPath]];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];

    // Configure the cell...
    cell = [DynamicTypeHelper setupDynamicTypeCell:cell style:UITableViewCellStyleSubtitle];
    cell.textLabel.text = [self textLabelText:indexPath];
    cell.detailTextLabel.text = [self detailTextLabelText:indexPath];

    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    switch (self.selectedAddress) {
        case SelectedAddressState: {
            
            SelectViewController *viewController = [self.storyboard instantiateViewControllerWithIdentifier:@"SelectViewController"];
            viewController.selectedAddress = SelectedAddressCityTown;
            viewController.selectedState = self.objects[indexPath.section][indexPath.row];
            viewController.title = self.objects[indexPath.section][indexPath.row];
            [self.navigationController pushViewController:viewController animated:YES];
            
            break;
        }
        case SelectedAddressCityTown: {

            PostalCodeModel *model = self.objects[indexPath.section];

            SelectViewController *viewController = [self.storyboard instantiateViewControllerWithIdentifier:@"SelectViewController"];
            viewController.selectedAddress = SelectedAddressStreet;
            viewController.selectedState = self.selectedState;
            viewController.selectedCityTown = model.cityTownK;
            viewController.title = model.cityTownK;
            [self.navigationController pushViewController:viewController animated:YES];

            break;
        }
        case SelectedAddressStreet: {

            PostalCodeModel *model = self.objects[indexPath.section][indexPath.row];
            DetailViewController *viewController = [self.storyboard instantiateViewControllerWithIdentifier:@"DetailViewController"];
            viewController.postalCodeModel = model;
            [self.navigationController pushViewController:viewController animated:YES];

            break;
        }
    }
}

@end

File: PostalCode/ViewController/SelectViewController.swift
//
//  SelectViewController.swift
//  PostalCode
//
//  Created by Koichiro OCHIISHI on 2025/05/05.
//  Copyright © 2025 OCHIISHI Koichiro. All rights reserved.
//

import UIKit

class SelectViewController: BaseTableViewController {

    enum SelectedAddress: Int {
        case state = 0  // 都道府県
        case cityTown = 1  // 市町村
        case street = 2  // 区群
    }

    var selectedAddress: SelectedAddress = .state
    var selectedState: String?
    var selectedCityTown: String?

    private var objects: [Any] = []
    private var sectionIndexTitles: [String] = []

    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.estimatedRowHeight = 44.0
        tableView.rowHeight = UITableView.automaticDimension
        tableView.sectionIndexTrackingBackgroundColor = Color.sectionIndexBackground

        if selectedAddress != .state {
            navigationItem.leftBarButtonItem = nil
        }

        (navigationController as? BaseNavigationController)?.startLoading()

        Task {
            do {
                switch selectedAddress {
                case .state:
                    let (fetchedObjects, fetchedSectionIndexTitles) = try await fetchStateData()
                    self.objects = fetchedObjects
                    self.sectionIndexTitles = fetchedSectionIndexTitles
                case .cityTown:
                    guard let selectedState = selectedState else { return }
                    let (fetchedObjects, fetchedSectionIndexTitles) = try await fetchCityTownData(
                        selectedState: selectedState)
                    self.objects = fetchedObjects
                    self.sectionIndexTitles = fetchedSectionIndexTitles
                case .street:
                    guard let selectedState = selectedState, let selectedCityTown = selectedCityTown
                    else { return }
                    let (fetchedObjects, fetchedSectionIndexTitles) = try await fetchStreetData(
                        selectedState: selectedState, selectedCityTown: selectedCityTown)
                    self.objects = fetchedObjects
                    self.sectionIndexTitles = fetchedSectionIndexTitles
                }
                (self.navigationController as? BaseNavigationController)?.stopLoading()
                self.tableView.reloadData()
            } catch {
                (self.navigationController as? BaseNavigationController)?.stopLoading()
                print("Error fetching data: \(error)")
            }
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        // 「読み込み中」が表示された状態で「戻る」が押された場合に、「読み込み中」が表示され続けてしまう問題を修正
        if navigationController?.viewControllers.firstIndex(of: self) == nil {
            (navigationController as? BaseNavigationController)?.stopLoading()
        }
    }

    // MARK: - Helper Methods

    private func textLabelText(for indexPath: IndexPath) -> String {
        switch selectedAddress {
        case .state:
            return (objects[indexPath.section] as? [String])?[indexPath.row] as? String ?? ""
        case .cityTown:
            let postalCode = objects[indexPath.section] as? PostalCode
            return postalCode?.cityTownK ?? ""
        case .street:
            let postalCode =
                (objects[indexPath.section] as? [PostalCode])?[indexPath.row]
                as? PostalCode
            return postalCode?.streetK.isEmpty ?? true ? postalCode?.cityTownK ?? "" : postalCode?.streetK ?? ""
        }
    }

    private func detailTextLabelText(for indexPath: IndexPath) -> String {
        switch selectedAddress {
        case .cityTown:
            let postalCode = objects[indexPath.section] as? PostalCode
            return postalCode?.cityTownH ?? ""
        case .street:
            let postalCode =
                (objects[indexPath.section] as? [PostalCode])?[indexPath.row]
                as? PostalCode
            return postalCode?.streetK.isEmpty ?? true ? postalCode?.cityTownH ?? "" : postalCode?.streetH ?? ""
        default:
            return ""
        }
    }

    private func fetchStateData() async throws -> ([[String]], [String]) {
        let states = await PostalCodeRepository.shared.getStates()
        let indexTitles = await PostalCodeRepository.shared.getStateSectionIndexTitles()
        return (states, indexTitles)
    }

    private func fetchCityTownData(selectedState: String) async throws -> (
        [PostalCode], [String]
    ) {
        let cityTowns = await PostalCodeRepository.shared.getCityTownsByState(selectedState)
        let indexTitles = cityTowns.map { postalCode in
            String(postalCode.cityTownK.prefix(3))
        }
        return (cityTowns, indexTitles)
    }

    private func fetchStreetData(selectedState: String, selectedCityTown: String) async throws -> (
        [[PostalCode]], [String]
    ) {
        let streets = await PostalCodeRepository.shared.getStreetsByState(
            selectedState, byCityAndTown: selectedCityTown)
        let indexTitles = streets.map { postalCodes in
            guard let postalCode = postalCodes.last else { return "" }
            return postalCode.streetH.isEmpty ? "" : String(postalCode.streetH.prefix(1))
        }
        return (streets, indexTitles)
    }

    // MARK: - Table View Data Source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return objects.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if selectedAddress == .cityTown {
            return 1
        }
        return (objects[section] as? [Any])?.count ?? 0
    }

    override func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return sectionIndexTitles
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int)
        -> String?
    {
        if selectedAddress == .state {
            return sectionIndexTitles[section]
        }
        return nil
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)
        -> UITableViewCell
    {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = textLabelText(for: indexPath)
        cell.detailTextLabel?.text = detailTextLabelText(for: indexPath)
        return cell
    }

    // MARK: - Table View Delegate

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        switch selectedAddress {
        case .state:
            let viewController =
                storyboard?.instantiateViewController(withIdentifier: "SelectViewController")
                as! SelectViewController
            viewController.selectedAddress = .cityTown
            viewController.selectedState =
                (objects[indexPath.section] as? [String])?[indexPath.row] as? String ?? ""
            viewController.title =
                (objects[indexPath.section] as? [String])?[indexPath.row] as? String ?? ""
            navigationController?.pushViewController(viewController, animated: true)
        case .cityTown:
            let postalCode = objects[indexPath.section] as? PostalCode
            let viewController =
                storyboard?.instantiateViewController(withIdentifier: "SelectViewController")
                as! SelectViewController
            viewController.selectedAddress = .street
            viewController.selectedState = selectedState
            viewController.selectedCityTown = postalCode?.cityTownK
            viewController.title = postalCode?.cityTownK
            navigationController?.pushViewController(viewController, animated: true)
        case .street:
            let postalCode =
                (objects[indexPath.section] as? [PostalCode])?[indexPath.row]
                as? PostalCode
            let viewController =
                storyboard?.instantiateViewController(withIdentifier: "DetailViewController")
                as! DetailViewController
            viewController.postalCode = postalCode
            navigationController?.pushViewController(viewController, animated: true)
        }
    }
}
